### ISSUE TITLE
Deadlock: trying to open a store while a writer is in progress

### ISSUE THREAD

**Issue by ncloudioj** [mozilla/rkv#59]
Ran into this when I was testing #58, following snippet can reproduce the deadlock.

```rs
let store = rkv.open_or_create("store");
let writer = store.write();
let another_store = rkv.open_or_create("another_store");
```

The deadlock was caused by the fact that LMDB doesn't allow two write transactions created at the same time. In this case, the `writer` is a wrapper of `RwTransaction`. Under the hood, `open_or_create` will create another write transaction to open a store, it'll hang forever since `writer` will never get a chance to abort or commit. So the solution would be either open all the stores before spawning any writer, or commit the writer and then open other stores.

We should at least document this for `Store` to avoid this kind of deadlock.

**Comment by mykmelez** [comment]
Nice catch!

We do currently document the potential for a call to _Store::write()_ to block on an extant Writer (in the [Basic Usage example](https://docs.rs/rkv/0.3.2/rkv/#basic-usage) and the [Store::write() API reference](https://docs.rs/rkv/0.3.2/rkv/struct.Store.html#method.write)). However, we don't currently document that _Rkv::open_or_create()_ will also block on an extant Writer, given the details of the underlying implementation. And that'll trip up API consumers. So we should definitely document it (and perhaps research ways to fix or work around the limitation, if it's feasible to do so).

**Comment by ncloudioj** [comment]
This also reminds me that we should also allow a store to be opened in the read_only mode. Doing so allows the consumer to open the store to read while there is an in-flight write transaction on the same store.

This might require some changes in lmdb-rs.

**Comment by mykmelez** [comment]
> This might require some changes in lmdb-rs.

Indeed. I've not had luck engaging with the lmdb-rs project recently. I have several PRs pending, the first two of which are now two months old. And you've received no response to your issue https://github.com/danburkert/lmdb-rs/issues/33. So I think we're going to need to fork the project and maintain our own fork, at least for the time being, in order to make this and other changes to it.

**Comment by ncloudioj** [comment]
Agreed, let's fork it for now. Its code looks pretty clean and well-organized, should be easy for us to maintain and develop.

**Comment by mykmelez** [comment]
> Agreed, let's fork it for now. Its code looks pretty neat and well-organized, should be easy for us to maintain and develop.

Ok, I've forked it to https://github.com/mozilla/lmdb-rs/ and added you as a collaborator!
